### DISCUSSION TITLE
what are item IDs used for?

### DISCUSSION THREAD

**Discussion by ghost** [flareteam/flare-game#931] [General]
[edit]
Hopefully this is the right place to ask for this, I only followed the "forum" link on flarerpg...

Hello.
I was looking at flare's code to see if I could give a try at implementing something like a PoC for #1815, and when looking at actual data, I noticed items have an ID (which is even a string...).
This ID seems to only be used for sets, yet it seems to be mandatory for all items, which means a modder *have* to manually maintain that. This is a quite weird situation if you ask me, so I'm asking if there's other roles for this field.

If not, it could probably be possible to automatically generate IDs, removing the maintenance constraint. Automatically generated IDs could have a prefix, to keep compat with the current feature (for sets). And thus, it would (probably) be rather easy to bake a mechanism in flare that would allow generating ranges of items.
As to define a range, well, I see 4 parameters:

* starting value
* number of steps
* modifier value
* modifier function (like, addition, subtraction, peeking a value from a defined list, exponents, etc)

so that in the end, the game would read a single section, and generate a bunch of items from it, at parse time.
This would not scale automatically with the character, sure, but it would be much easier to generate ranges of monsters or items (without relying on external tooling for nothing).
This problem could probably be solved by having conditions that allows or prevents apparition of a specific thing (being enemy, NPC, item loot, item in a merchant's stash... I don't know).
Note that I do not really know how flare's modding work, so maybe my question and/or idea is stupid. I'm just looking at the code to check if I can hack a thing or two in there, is all.

A quick example would be:

```
[item]
id=shirt$
name=Cloth Shirt
INCLUDE items/base/armor/cloth/chest.txt
level=$1,4,1,addition
price=$75,4,1.4,multiply
```

which would result in 4 "cloth shirts":

* shirt1, level 1, price 75
* shirt2, level 2, price 90
* shirt3, level 3, price 108
* shirt4, level 4, price 129.6

Well, that's just some quick thinking.

**Comment by dorkster** [comment]
- Item IDs are not strings, but unsigned integers (specifically, `size_t`).
  - If you're wondering why they're *not* strings, item IDs were originally just indexes into an array/`std::vector`. They now function as keys into a `std::map`, but the mod API has remained the same. It's just never made sense to change them to strings.
- They're used for more than just item sets. Try searching the [Attribute Reference](https://github.com/flareteam/flare-engine/wiki/Attribute-Reference) page for `item_id`.

Not entirely sold on the idea of auto-generating IDs, but it's an interesting idea...

**Reply by ghost** [reply]
> Item IDs are not strings

Weird, this is not what the source code seems to say: https://github.com/flareteam/flare-engine/blob/master/src/ItemManager.h#L197-L207
Wiki does not specifies the underlying type, only says it's a unique identifier.

> Try searching the Attribute Reference page for item_id

Interesting, grepping the codebase gives no result for the string `"item_id"`

> Not entirely sold on the idea of auto-generating IDs, but it's an interesting idea...

That's just an idea I had after a quick thinking around, nothing more. And the easier way I see to build stuff that requires less maintenance while not requiring to modify too many things in the codebase: generating list of items from a single description means only changing the parser, at the cost of increase memory usage and starting time, of course.

**Reply by dorkster** [reply]
>Weird, this is not what the source code seems to say: https://github.com/flareteam/flare-engine/blob/master/src/ItemManager.h#L197-L207

That ID is for **ItemType**, not Item. ItemType identifiers are indeed strings. An example configuration file looks like this: https://github.com/flareteam/flare-game/blob/master/mods/fantasycore/items/types.txt

Since Item IDs are keys in a std::map, they don't need to be part of the Item class itself. Here's where the parsing of Item IDs occurs: https://github.com/flareteam/flare-engine/blob/master/src/ItemManager.cpp#L137-L140

>Interesting, grepping the codebase gives no result for the string "item_id"

The `item_id` moniker is more for documentation. In the code, it's just a typedef: https://github.com/flareteam/flare-engine/blob/master/src/Utils.h#L40

That said, here's what you'll see when running `grep -rn item_id | grep ATTR | sort` in the `src/` directory (searching for "ATTR" singles out all the documentation comments):

```
EngineSettings.cpp:121:                         // @ATTR currency_id|item_id|An item id that will be used as currency.
EngineSettings.cpp:574:                 // @ATTR equipment|list(item_id)|A list of items that are equipped when starting with this class.
EngineSettings.cpp:576:                 // @ATTR carried|list(item_id)|A list of items that are placed in the normal inventorty when starting with this class.
EventManager.cpp:412:           // @ATTR event.requires_item|list(item_id)|Event requires specific item (not equipped). Quantity can be specified by appending ":Q" to the item_id, where Q is an integer.
EventManager.cpp:435:           // @ATTR event.requires_not_item|list(item_id)|Event requires not having a specific item (not equipped). Quantity can be specified by appending ":Q" to the item_id, where Q is an integer.
EventManager.cpp:514:           // @ATTR event.remove_item|list(item_id)|Removes specified item from hero inventory. Quantity can be specified by appending ":Q" to the item_id, where Q is an integer.
EventManager.cpp:549:           // @ATTR event.reward_item|(list(item_id)|Reward hero with a specified item. Quantity can be specified by appending ":Q" to the item_id, where Q is an integer. To maintain backwards compatibility, the quantity must be defined for at least the first item in the list in order to use this syntax.
EventManager.cpp:550:           // @ATTR event.reward_item|item_id, int : Item, Quantity|Reward hero with y number of item x. NOTE: This syntax is maintained for backwards compatibility. It is recommended to use the above syntax instead.
ItemManager.cpp:138:                    // @ATTR id|item_id|An uniq id of the item used as reference from other classes.
ItemManager.cpp:556:                    // @ATTR items|list(item_id)|List of item id's that is part of the set.
LootManager.cpp:664:                            // @ATTR loot.id|[item_id, "currency"]|The ID of the loot item. "currency" will use the item ID defined as currency_id in engine/misc.txt.
Map.cpp:411:            // @ATTR enemygroup.requires_item|list(item_id)|Item required to exist in player inventory to load enemy group. Quantity can be specified by appending ":Q" to the item_id, where Q is an integer.
Map.cpp:423:            // @ATTR enemygroup.requires_not_item|list(item_id)|Item required to not exist in player inventory to load enemy group. Quantity can be specified by appending ":Q" to the item_id, where Q is an integer.
Map.cpp:565:            // @ATTR npc.requires_item|list(item_id)|Item required to exist in player inventory to load NPC. Quantity can be specified by appending ":Q" to the item_id, where Q is an integer.
Map.cpp:577:            // @ATTR npc.requires_not_item|list(item_id)|Item required to not exist in player inventory to load NPC. Quantity can be specified by appending ":Q" to the item_id, where Q is an integer.
NPC.cpp:223:                                    // @ATTR npc.constant_stock|repeatable(list(item_id))|A list of items this vendor has for sale. Quantity can be specified by appending ":Q" to the item_id, where Q is an integer.
NPC.cpp:230:                                    // @ATTR npc.status_stock|repeatable(string, list(item_id)) : Required status, Item(s)|A list of items this vendor will have for sale if the required status is met. Quantity can be specified by appending ":Q" to the item_id, where Q is an integer.
PowerManager.cpp:427:                   // @ATTR power.requires_item|repeatable(item_id, int) : Item, Quantity|Requires a specific item of a specific quantity in inventory. If quantity > 0, then the item will be removed.
PowerManager.cpp:435:                   // @ATTR power.requires_equipped_item|repeatable(item_id, int) : Item, Quantity|Requires a specific item of a specific quantity to be equipped on hero. If quantity > 0, then the item will be removed.
QuestLog.cpp:125:                       // @ATTR quest.requires_item|list(item_id)|Quest requires specific item (not equipped)
QuestLog.cpp:129:                       // @ATTR quest.requires_not_item|list(item_id)|Quest requires not having a specific item (not equipped)
StatBlock.cpp:671:              // @ATTR first_defeat_loot|item_id|Drops this item upon first death.
StatBlock.cpp:673:              // @ATTR quest_loot|string, string, item_id : Required status, Required not status, Item|Drops this item when campaign status is met.
```

**Reply by ghost** [reply]
So, what is the difference between an **itemType** and an **item**? I would tend to think that an item is an instance of an itemType, with a few additional information such at location, maybe owner, that kind of things.

If so, then I've been using the wrong terms all along: I was talking about generating **itemType**s and not **item**s, from a skeleton.
I still need to know what ItemType IDs are for, though, other than the sets. And since they're strings, the idea of generating those would require some kind of reserved prefix to prevent collisions, and this could even be extended to get a simple "namespace" mechanism, so that generated item types could still be parts of sets (why not after all).

Still quickly-thinking, I would probably do something like having the `id` key optional. When optional, it would generate an ID like `AUTO::[integer]` where `AUTO:` would be a reserved prefix, `:` would be separator between fields, and `[integer]` would be the generated part (incremental thing probably). When the `id` key would be present but does not starts with `AUTO:`, everything would work as they currently do. When it does, though, then it would be something like that: `id=AUTO:foobar`, then the generated ID would look like `id=AUTO:foobar:[integer]`. The idea here would be to able to refer to randomly generated stuff from sets or whatever could need it. Some kind of namespace system if you prefer.
I'm not sure how much this would be doable (or desirable?) though. Probably not that hard to implement, from what I've seen in the codebase, but I only had a quick look at it, one never knows.

**Reply by dorkster** [reply]
- Item IDs are a numerical identifier used wherever we need to reference an item (loot tables, inventory storage, etc)
- ItemType IDs are a string identifier used to refer to which equipment slot an item can be worn in.

```
# example
[item]
id=1 <-- Item ID
item_type=ring <-- Item Type ID
```

So, ItemType is irrelevant to the ideas you're proposing.

There are two distinct problems we potentially want to solve with items: scaling and randomization. Your example in the first post targets the former, and the example in your latest post is targets the latter. Here's why I don't think auto-generating Item IDs is good for either:

- Scaling: I simply would rather base this around item level instead of changing how IDs work. It's more similar to how we scale the player's stats across level-ups.
- Randomization: From your example, `id=AUTO:foobar:[integer]`, we'd be generating `[integer]` number of random permutations of `foobar` upfront. Unless, I'm misunderstanding, that seems very limiting unless you use a large number of permutations. And then in that case, you're wasting memory on items the player will most likely never see. That's why I prefer my potential approach that makes the items as needed with a base item ID, a level, and a seed value.

**Comment by ghost** [comment]
> and the example in your latest post is targets the latter

Then I failed to explain my point. I have not thought to address randomization of items, or rather not pure randomisation, but you're right that this is important to consider.

> Unless, I'm misunderstanding

You did understood correctly. And yes, the mechanism I'm talking about have the limitation that it requires generating things upfront. The "randomisation" would only happen because it would be easier to create a bigger "loot stash" (number of items you might possibly loot).

> That's why I prefer my potential approach that makes the items as needed with a base item ID, a level, and a seed value.

I agree it would be superior, but it also seems more complex to implement. But then, I don't know the codebase much, so most stuff would require time from me to implement :)
This is why I preferred to use forum instead of ticket to talk about that.
